File: pkg/dart2cpp/sample/dart/hello_test_enhanced.dart
/// Dart 增强语法测试示例
///
/// 包含更多高级Dart语法特性，用于测试Dart到C++的转换
///

void main() {
  print('🔥 Dart 增强语法测试开始');

  // 1. 基本数据类型
  testBasicTypes();

  // 2. 控制流
  testControlFlow();

  // 3. 函数（增强版）
  testEnhancedFunctions();

  // 4. 集合类型（增强版）
  testEnhancedCollections();

  // 5. 类和对象（增强版）
  testEnhancedClasses();

  // 6. 泛型
  testGenerics();

  // 7. 异常处理
  testExceptions();

  // 8. 空安全特性
  testNullSafety();

  print('✅ Dart 增强语法测试完成');
}

/// 测试基本数据类型
void testBasicTypes() {
  print('\n--- 基本数据类型 ---');

  // 数值类型
  int age = 25;
  double price = 99.99;
  int quantity = 10;

  // 字符串类型
  String name = '张三';
  String message = '你好，世界！';

  // 布尔类型
  bool isStudent = true;
  bool isEmployed = false;

  // 字符串操作
  String upperMessage = message.toUpperCase();
  String lowerMessage = message.toLowerCase();
  int messageLength = message.length;

  print('基本类型操作完成');
}

/// 测试控制流
void testControlFlow() {
  print('\n--- 控制流 ---');

  // if-else 语句
  int score = 85;
  if (score >= 90) {
    print('优秀');
  } else if (score >= 80) {
    print('良好');
  } else if (score >= 60) {
    print('及格');
  } else {
    print('不及格');
  }

  // switch 语句
  String grade = 'B';
  switch (grade) {
    case 'A':
      print('优秀');
      break;
    case 'B':
      print('良好');
      break;
    case 'C':
      print('及格');
      break;
    default:
      print('未知等级');
  }

  // for 循环
  print('For 循环:');
  for (int i = 0; i < 5; i++) {
    print('数字: $i');
  }

  // for-in 循环
  List<String> fruits = ['苹果', '香蕉', '橙子'];
  print('For-in 循环:');
  for (String fruit in fruits) {
    print('水果: $fruit');
  }

  // while 循环
  print('While 循环:');
  int count = 3;
  while (count > 0) {
    print('倒计时: $count');
    count--;
  }

  // do-while 循环
  print('Do-while 循环:');
  int number = 1;
  do {
    print('数字: $number');
    number *= 2;
  } while (number <= 10);

  // 条件表达式
  int testAge = 25;
  bool isAdult = testAge >= 18 ? true : false;
  print('成年人: $isAdult');
}

/// 测试增强函数
void testEnhancedFunctions() {
  print('\n--- 增强函数 ---');

  // 带默认参数的函数
  String greet(String name, [String greeting = '你好']) {
    return '$greeting, $name!';
  }

  // 带命名参数的函数
  String createProfile({required String name, int? age, String? city}) {
    String profile = '姓名: $name';
    if (age != null) {
      profile += ', 年龄: $age';
    }
    if (city != null) {
      profile += ', 城市: $city';
    }
    return profile;
  }

  // 高阶函数
  int calculate(int a, int b, int Function(int, int) operation) {
    return operation(a, b);
  }

  // 箭头函数
  int add(int a, int b) => a + b;
  int multiply(int a, int b) => a * b;

  // 调用函数
  print(greet('李四'));
  print(greet('王五', '早上好'));
  print(createProfile(name: '赵六', age: 30, city: '北京'));

  int sum = calculate(5, 3, add);
  int product = calculate(5, 3, multiply);
  print('计算结果: 和=$sum, 积=$product');
}

/// 测试增强集合
void testEnhancedCollections() {
  print('\n--- 增强集合 ---');

  // List 操作
  List<int> numbers = [1, 2, 3, 4, 5];
  List<String> names = ['张三', '李四', '王五'];

  // Map 操作
  Map<String, int> scores = {'张三': 95, '李四': 87, '王五': 92};

  // Set 操作
  Set<String> uniqueNames = {'张三', '李四', '王五'};

  // 集合方法
  print('数字列表长度: ${numbers.length}');
  print('第一个数字: ${numbers.first}');
  print('最后一个数字: ${numbers.last}');

  // 列表变换
  List<int> doubled = numbers.map((n) => n * 2).toList();
  List<int> evens = numbers.where((n) => n % 2 == 0).toList();
  int total = numbers.reduce((a, b) => a + b);

  print('翻倍后的列表: $doubled');
  print('偶数列表: $evens');
  print('数字总和: $total');

  // Map 操作
  print('张三的分数: ${scores['张三']}');
  scores['赵六'] = 88;
  print('更新后的分数: $scores');
}

/// 测试增强类
void testEnhancedClasses() {
  print('\n--- 增强类 ---');

  // 创建对象
  Person person = Person('张三', 25);
  print('人员信息: ${person.toString()}');

  // 使用静态方法
  Person.logCreation();

  // 使用 getter 和 setter
  print('人员姓名: ${person.name}');
  person.name = '李四';
  print('更新后姓名: ${person.name}');

  // 使用抽象类
  Shape circle = Circle(5.0);
  Shape rectangle = Rectangle(4.0, 6.0);

  print('圆形面积: ${circle.area()}');
  print('矩形面积: ${rectangle.area()}');

  // 使用 mixin
  Bird bird = Bird('鹦鹉');
  bird.fly();
  bird.speak();
}

/// 测试泛型
void testGenerics() {
  print('\n--- 泛型 ---');

  // 泛型类
  Box<String> stringBox = Box<String>('Hello');
  Box<int> intBox = Box<int>(42);

  print('字符串盒子内容: ${stringBox.getValue()}');
  print('整数盒子内容: ${intBox.getValue()}');

  // 泛型函数
  List<int> numbers = [1, 2, 3, 4, 5];
  List<String> strings = ['a', 'b', 'c'];

  int firstNumber = getFirst<int>(numbers);
  String firstString = getFirst<String>(strings);

  print('第一个数字: $firstNumber');
  print('第一个字符串: $firstString');
}

/// 测试异常处理
void testExceptions() {
  print('\n--- 异常处理 ---');

  try {
    int result = 10 ~/ 2; // 整数除法
    print('结果: $result');
  } catch (e) {
    print('捕获到异常: $e');
  }

  try {
    validateAge(-5);
  } catch (e) {
    print('年龄验证失败: $e');
  }
}

/// 测试空安全特性
void testNullSafety() {
  print('\n--- 空安全特性 ---');

  // 可空类型
  String? nullableString = 'Hello';
  String? nullString = null;

  // 空值合并操作符
  String nonNull1 = nullableString ?? 'default';
  String nonNull2 = nullString ?? 'default';

  print('空值合并结果: $nonNull1, $nonNull2');

  // 空值安全访问
  List<String>? nullableList = ['a', 'b', 'c'];
  int? length1 = nullableList?.length;
  print('列表长度: $length1');

  // 空值断言
  String definitelyNotNull = nullableString!;
  print('断言非空值: $definitelyNotNull');
}

// ========================
// 增强类定义
// ========================

/// 人员类（增强版）
class Person {
  String _name;
  int _age;

  Person(this._name, this._age);

  // Getter 和 Setter
  String get name => _name;
  set name(String value) => _name = value;

  int get age => _age;
  set age(int value) => _age = value;

  /// 自我介绍
  void introduce() {
    print('大家好，我是 $_name，今年 $_age 岁。');
  }

  /// 静态方法
  static void logCreation() {
    print('创建了一个新的Person实例');
  }

  @override
  String toString() {
    return 'Person{name: $_name, age: $_age}';
  }
}

/// 抽象形状类
abstract class Shape {
  double area();
}

/// 圆形类
class Circle extends Shape {
  double radius;

  Circle(this.radius);

  @override
  double area() {
    return 3.14159 * radius * radius;
  }
}

/// 矩形类
class Rectangle extends Shape {
  double width;
  double height;

  Rectangle(this.width, this.height);

  @override
  double area() {
    return width * height;
  }
}

/// 飞行能力 Mixin
mixin Flyable {
  void fly() {
    print('正在飞行');
  }
}

/// 说话能力 Mixin
mixin Speakable {
  void speak() {
    print('正在说话');
  }
}

/// 鸟类（使用 Mixin）
class Bird with Flyable, Speakable {
  String name;

  Bird(this.name);
}

// ========================
// 泛型定义
// ========================

/// 泛型盒子类
class Box<T> {
  T _value;

  Box(this._value);

  T getValue() => _value;
  void setValue(T value) => _value = value;
}

/// 泛型函数
T getFirst<T>(List<T> list) {
  if (list.isEmpty) {
    throw Exception('列表为空');
  }
  return list[0];
}

// ========================
// 工具函数
// ========================

/// 验证年龄的函数
void validateAge(int age) {
  if (age < 0) {
    throw Exception('年龄不能为负数');
  }
  print('有效年龄: $age');
}

File: pkg/dart2cpp/output/hello_test_enhanced.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: Person
// ============================================================================

class Person {
private:
  String _name;
  Int _age;
public:
  Person(String _name, Int _age) : _name(_name), _age(_age) {
  }
  
  String name() {
    return this->_name;
  }
  
  Nullable name(String value) {
    return this->_name = value;
  }
  
  Int age() {
    return this->_age;
  }
  
  Nullable age(Int value) {
    return this->_age = value;
  }
  
  Nullable introduce() {
    dart_print(dart_concat(dart_string("大家好，我是 "), (this->_name).toString(), dart_string("，今年 "), (this->_age).toString(), dart_string(" 岁。")));
return Void;
  }
  
  Nullable logCreation() {
    dart_print(dart_string("创建了一个新的Person实例"));
return Void;
  }
  
  String toString() {
    return dart_concat(dart_string("Person{name: "), (this->_name).toString(), dart_string(", age: "), (this->_age).toString(), dart_string("}"));
  }
  
};

// ============================================================================
// 类: Shape
// ============================================================================

DART_INTERFACE(Shape)
  DART_ABSTRACT_METHOD(Double, area, ())
DART_INTERFACE_END

// ============================================================================
// 类: Circle
// ============================================================================

class Circle : public Shape {
public:
  Double radius;
  Circle(Double radius) : radius(radius) {
  }
  
  Double area() {
    return dart_double(3.14159)->operator_mul(this->radius)->operator_mul(this->radius);
  }
  
};

// ============================================================================
// 类: Rectangle
// ============================================================================

class Rectangle : public Shape {
public:
  Double width;
  Double height;
  Rectangle(Double width, Double height) : width(width), height(height) {
  }
  
  Double area() {
    return this->width->operator_mul(this->height);
  }
  
};

// ============================================================================
// 类: Flyable
// ============================================================================

DART_INTERFACE(Flyable)
DART_INTERFACE_END

// ============================================================================
// 类: Speakable
// ============================================================================

DART_INTERFACE(Speakable)
DART_INTERFACE_END

// ============================================================================
// 类: Bird
// ============================================================================

class Bird : public _Bird&Object&Flyable&Speakable {
public:
  String name;
  Bird(String name) : name(name) {
  }
  
};

// ============================================================================
// 类: Box
// ============================================================================

template<typename T>

class Box {
private:
  T _value;
public:
  Box(T _value) : _value(_value) {
  }
  
  T getValue() {
    return this->_value;
  }
  
  Nullable setValue(T value) {
    return this->_value = value;
  }
  
};

// ============================================================================
// 类: _Bird&Object&Flyable
// ============================================================================

DART_INTERFACE(_Bird&Object&Flyable)
DART_INTERFACE_END

// ============================================================================
// 类: _Bird&Object&Flyable&Speakable
// ============================================================================

DART_INTERFACE(_Bird&Object&Flyable&Speakable)
DART_INTERFACE_END

Nullable testBasicTypes();
Nullable testControlFlow();
Nullable testEnhancedFunctions();
Nullable testEnhancedCollections();
Nullable testEnhancedClasses();
Nullable testGenerics();
Nullable testExceptions();
Nullable testNullSafety();
template<typename T>
T getFirst(ObjectPtr<List<T>> list);
Nullable validateAge(Int age);
Nullable testBasicTypes() {
  dart_print(dart_string("\n--- 基本数据类型 ---"));
auto age = dart_int(25);
auto price = dart_double(99.99);
auto quantity = dart_int(10);
auto name = dart_string("张三");
auto message = dart_string("你好，世界！");
auto isStudent = dart_bool(true);
auto isEmployed = dart_bool(false);
auto upperMessage = message->toUpperCase();
auto lowerMessage = message->toLowerCase();
auto messageLength = message->size();
dart_print(dart_string("基本类型操作完成"));
return Void;
}

Nullable testControlFlow() {
  dart_print(dart_string("\n--- 控制流 ---"));
auto score = dart_int(85);
if (score->operator_greater_equals(dart_int(90))) {
dart_print(dart_string("优秀"));
} else {
if (score->operator_greater_equals(dart_int(80))) {
dart_print(dart_string("良好"));
} else {
if (score->operator_greater_equals(dart_int(60))) {
dart_print(dart_string("及格"));
} else {
dart_print(dart_string("不及格"));
}
}
}
auto grade = dart_string("B");
if (grade == dart_string("A")) {
  dart_print(dart_string("优秀"));
} else if (grade == dart_string("B")) {
  dart_print(dart_string("良好"));
} else if (grade == dart_string("C")) {
  dart_print(dart_string("及格"));
}else {
  dart_print(dart_string("未知等级"));
}
dart_print(dart_string("For 循环:"));
for (auto i = dart_int(0); i->operator_less(dart_int(5)); i = i->operator_add(dart_int(1))) {
dart_print(dart_string("数字: ") + (i).toString());
}
auto fruits = dart_literal(dart_string("苹果"), dart_string("香蕉"), dart_string("橙子"));
dart_print(dart_string("For-in 循环:"));
auto sync_for_iterator = fruits->iterator();
for (; sync_for_iterator->hasNext(); ) {
auto fruit = sync_for_iterator->next();
dart_print(dart_string("水果: ") + (fruit).toString());
}
dart_print(dart_string("While 循环:"));
auto count = dart_int(3);
while (count->operator_greater(dart_int(0))) {
dart_print(dart_string("倒计时: ") + (count).toString());
count = count->operator_sub(dart_int(1));
}
dart_print(dart_string("Do-while 循环:"));
auto number = dart_int(1);
do {
dart_print(dart_string("数字: ") + (number).toString());
number = number->operator_mul(dart_int(2));
} while ((number->operator_less_equals(dart_int(10))).toBool());
auto testAge = dart_int(25);
auto isAdult = testAge->operator_greater_equals(dart_int(18)) ? dart_bool(true) : dart_bool(false);
dart_print(dart_string("成年人: ") + (isAdult).toString());
return Void;
}

Nullable testEnhancedFunctions() {
  dart_print(dart_string("\n--- 增强函数 ---"));
auto greet = [&](String name, String greeting) -> String return dart_concat((greeting).toString(), dart_string(", "), (name).toString(), dart_string("!"));;
auto createProfile = [&]() -> String auto profile = dart_string("姓名: ") + (name).toString();
if (!(dart_is_null(age))) {
profile = profile->operator_add(dart_string(", 年龄: ") + (age).toString());
}
if (!(dart_is_null(city))) {
profile = profile->operator_add(dart_string(", 城市: ") + (city).toString());
}
return profile;;
auto calculate = [&](Int a, Int b, ObjectPtr<TypedFunction<std::function<Int(Int, Int)>, Int, Int, Int>> operation) -> Int return operation->call(a, b);;
auto add = [&](Int a, Int b) -> Int return a->operator_add(b);;
auto multiply = [&](Int a, Int b) -> Int return a->operator_mul(b);;
dart_print(greet(dart_string("李四")));
dart_print(greet(dart_string("王五"), dart_string("早上好")));
dart_print(createProfile(/*name:*/ dart_string("赵六"), /*age:*/ dart_int(30), /*city:*/ dart_string("北京")));
auto sum = calculate(dart_int(5), dart_int(3), add);
auto product = calculate(dart_int(5), dart_int(3), multiply);
dart_print(dart_concat(dart_string("计算结果: 和="), (sum).toString(), dart_string(", 积="), (product).toString()));
return Void;
}

Nullable testEnhancedCollections() {
  dart_print(dart_string("\n--- 增强集合 ---"));
auto numbers = dart_literal(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
auto names = dart_literal(dart_string("张三"), dart_string("李四"), dart_string("王五"));
auto scores = Map<String, Int>::createFromEntries({{dart_string("张三"), dart_int(95)}, {dart_string("李四"), dart_int(87)}, {dart_string("王五"), dart_int(92)}});
auto uniqueNames = ([&]() { const auto unnamed_var = ObjectPtr<Set>(new Set()); unnamed_var->add(dart_string("张三")); unnamed_var->add(dart_string("李四")); unnamed_var->add(dart_string("王五")); return unnamed_var; })();
dart_print(dart_string("数字列表长度: ") + (numbers->size()).toString());
dart_print(dart_string("第一个数字: ") + (numbers->first()).toString());
dart_print(dart_string("最后一个数字: ") + (numbers->last()).toString());
auto doubled = numbers->map(makeFunction([&](Int n) { return n->operator_mul(dart_int(2)); }))->toList();
auto evens = numbers->where(makeFunction([&](Int n) { return (n->operator_mod(dart_int(2)) == dart_int(0)); }))->toList();
auto total = numbers->reduce(makeFunction([&](Int a, Int b) { return a->operator_add(b); }));
dart_print(dart_string("翻倍后的列表: ") + (doubled).toString());
dart_print(dart_string("偶数列表: ") + (evens).toString());
dart_print(dart_string("数字总和: ") + (total).toString());
dart_print(dart_string("张三的分数: ") + (scores->operator_index(dart_string("张三"))).toString());
scores->operator_index_set(dart_string("赵六"), dart_int(88));
dart_print(dart_string("更新后的分数: ") + (scores).toString());
return Void;
}

Nullable testEnhancedClasses() {
  dart_print(dart_string("\n--- 增强类 ---"));
auto person = ObjectPtr<Person>(new Person(dart_string("张三"), dart_int(25)));
dart_print(dart_string("人员信息: ") + (person->toString()).toString());
Person::logCreation();
dart_print(dart_string("人员姓名: ") + (person->name()).toString());
person->set_name(dart_string("李四"));
dart_print(dart_string("更新后姓名: ") + (person->name()).toString());
auto circle = ObjectPtr<Circle>(new Circle(dart_double(5.0)));
auto rectangle = ObjectPtr<Rectangle>(new Rectangle(dart_double(4.0), dart_double(6.0)));
dart_print(dart_string("圆形面积: ") + (circle->area()).toString());
dart_print(dart_string("矩形面积: ") + (rectangle->area()).toString());
auto bird = ObjectPtr<Bird>(new Bird(dart_string("鹦鹉")));
bird->fly();
bird->speak();
return Void;
}

Nullable testGenerics() {
  dart_print(dart_string("\n--- 泛型 ---"));
auto stringBox = ObjectPtr<Box>(new Box(dart_string("Hello")));
auto intBox = ObjectPtr<Box>(new Box(dart_int(42)));
dart_print(dart_string("字符串盒子内容: ") + (stringBox->getValue()).toString());
dart_print(dart_string("整数盒子内容: ") + (intBox->getValue()).toString());
auto numbers = dart_literal(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
auto strings = dart_literal(dart_string("a"), dart_string("b"), dart_string("c"));
auto firstNumber = getFirst(numbers);
auto firstString = getFirst(strings);
dart_print(dart_string("第一个数字: ") + (firstNumber).toString());
dart_print(dart_string("第一个字符串: ") + (firstString).toString());
return Void;
}

Nullable testExceptions() {
  dart_print(dart_string("\n--- 异常处理 ---"));
try {
auto result = dart_int(10)->truncatingDivision(dart_int(2));
dart_print(dart_string("结果: ") + (result).toString());
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
try {
validateAge(dart_int(-5));
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern;
return Void;
}

Nullable testNullSafety() {
  dart_print(dart_string("\n--- 空安全特性 ---"));
auto nullableString = dart_string("Hello");
auto nullString = Null;
auto nonNull1 = dart_null_coalesce(nullableString, dart_string("default"));
auto nonNull2 = dart_null_coalesce(nullString, dart_string("default"));
dart_print(dart_concat(dart_string("空值合并结果: "), (nonNull1).toString(), dart_string(", "), (nonNull2).toString()));
auto nullableList = dart_literal(dart_string("a"), dart_string("b"), dart_string("c"));
auto length1 = dart_null_coalesce(nullableList, Null);
dart_print(dart_string("列表长度: ") + (length1).toString());
auto definitelyNotNull = nullableString;
dart_print(dart_string("断言非空值: ") + (definitelyNotNull).toString());
return Void;
}

template<typename T>
T getFirst(ObjectPtr<List<T>> list) {
  if (list->isEmpty()) {
throw DartException(Exception::(dart_string("列表为空")));
}
return list->operator_index(dart_int(0));
}

Nullable validateAge(Int age) {
  if (age->operator_less(dart_int(0))) {
throw DartException(Exception::(dart_string("年龄不能为负数")));
}
dart_print(dart_string("有效年龄: ") + (age).toString());
return Void;
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    dart_print(dart_string("🔥 Dart 增强语法测试开始"));
testBasicTypes();
testControlFlow();
testEnhancedFunctions();
testEnhancedCollections();
testEnhancedClasses();
testGenerics();
testExceptions();
testNullSafety();
dart_print(dart_string("✅ Dart 增强语法测试完成"));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
